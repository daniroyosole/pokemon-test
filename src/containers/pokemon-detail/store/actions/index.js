import {
    GET_DETAIL_REQUEST,
    GET_DETAIL_SUCCESS,
    GET_DETAIL_FAILURE,
  } from './actionTypes';
  
  export const getPokemonDetail = information => ({
    type: GET_DETAIL_REQUEST,
    payload: information,
  });
  
  export const getPokemonDetailSuccess = response => ({
    type: GET_DETAIL_SUCCESS,
    response,
  });
  
  export const getPokemonDetailFailure = error => ({
    type: GET_DETAIL_FAILURE,
    error,
  });
