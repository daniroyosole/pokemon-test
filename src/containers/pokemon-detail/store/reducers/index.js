import {
    GET_DETAIL_REQUEST,
    GET_DETAIL_SUCCESS,
    GET_DETAIL_FAILURE
  } from '../actions/actionTypes';
  
  const initialState = {
    detail: {},
    loading: false,
    error: false
  }
  
  const reducer = (state = initialState, action) => {
    const { type, response } = action;
    switch(type) {
      case GET_DETAIL_REQUEST:
        return {
          ...state,
          detail: {},
          loading: true,
          error: false
        }
      case GET_DETAIL_SUCCESS:
        return {
          ...state,
          detail: response,
          loading: false
        }
      case GET_DETAIL_FAILURE:
        return {
          ...state,
          detail: {},
          loading: false,
          error: true
        }
      default:
        return state;
    }
  }
  
  export default reducer;