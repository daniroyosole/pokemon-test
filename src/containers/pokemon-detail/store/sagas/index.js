import { put } from 'redux-saga/effects';
import { getPokemonDetails } from '@Services/pokemon';
import {
  getPokemonDetailSuccess,
  getPokemonDetailFailure,
} from '../actions';
import Cookies from 'universal-cookie';
import { COOKIES_DETAIL } from '@Utils/utils';

export function* getDetail({payload}) {
  try {
    const response = yield getData(payload);
    if (typeof response !== "undefined" && typeof response.id !== "undefined") {
      const cookies = new Cookies();
      cookies.set(COOKIES_DETAIL+payload, response, {path: '/', expires: getNewTime(60*24)});
      yield put(getPokemonDetailSuccess(response));
    } else {
      yield put(getDetailFailure({}));
    }
  } catch(err) {
    yield put(getPokemonDetailFailure(err));
  }
};


function getData(payload) {
  const cookies = new Cookies();
  const cookiesData = cookies.get(COOKIES_DETAIL+payload);
  if (typeof cookiesData !== "undefined" && typeof cookiesData.id !== "undefined") {
      return cookiesData;     
  } else {
    return getPokemonDetails(payload);
  }
}