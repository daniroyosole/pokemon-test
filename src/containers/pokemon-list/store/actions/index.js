import {
    GET_LIST_REQUEST,
    GET_LIST_SUCCESS,
    GET_LIST_FAILURE,
  } from './actionTypes';
  
  export const getPokemonList = information => ({
    type: GET_LIST_REQUEST,
    payload: information,
  });
  
  export const getPokemonListSuccess = response => ({
    type: GET_LIST_SUCCESS,
    response,
  });
  
  export const getPokemonListFailure = error => ({
    type: GET_LIST_FAILURE,
    error,
  });
