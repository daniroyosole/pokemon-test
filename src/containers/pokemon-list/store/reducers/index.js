import {
    GET_LIST_REQUEST,
    GET_LIST_SUCCESS,
    GET_LIST_FAILURE
  } from '../actions/actionTypes';
  import { COOKIES_LIST, getNewTime } from '@Utils/utils';
  import Cookies from 'universal-cookie';

  const initialState = {
    pageNumber: 0,
    pokemonList: []
  }
  
  const reducer = (state = initialState, action) => {
    const { type, response } = action;
    const cookies = new Cookies();

    switch(type) {
      case GET_LIST_REQUEST:
        return state
      case GET_LIST_SUCCESS:
        const shortList = response.map((item) => ({name: item.name}));
        const newList = state.pokemonList.concat(shortList);
        cookies.set(COOKIES_LIST+(state.pageNumber+1), shortList, {path: '/', expires: getNewTime(60*24)});
        return {
          ...state,
          pokemonList: newList,
          pageNumber: state.pageNumber+1
        }
      case GET_LIST_FAILURE:
        return {
          ...state,
          pokemonList: [],
          pageNumber: 0
        }
      default:
        return state;
    }
  }
  
  export default reducer;