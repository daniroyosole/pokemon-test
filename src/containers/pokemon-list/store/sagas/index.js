import { put } from 'redux-saga/effects';
import { COOKIES_LIST } from '@Utils/utils';
import { getPokemonList } from '@Services/pokemon';
import {
  getPokemonListSuccess,
  getPokemonListFailure,
} from '../actions';
import Cookies from 'universal-cookie';

export function* getList({payload}) {
  try {
    const response = yield getData(payload);
    if (typeof response !== "undefined" && typeof response.results !== "undefined") {
      yield put(getPokemonListSuccess(response.results));
    } else {
      yield put(getListFailure({}));
    }
  } catch(err) {
    yield put(getPokemonListFailure(err));
  }
};

function getData(payload) {
  const cookies = new Cookies();
  const cachedList = cookies.get(COOKIES_LIST+payload);
  if (typeof cachedList !== "undefined" && cachedList.length >= 0) {
    return { results: cachedList }; 
  } else {
    return getPokemonList(payload);
  }
}