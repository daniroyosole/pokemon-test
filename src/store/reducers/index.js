import { combineReducers } from 'redux';

// Container reducers
import pokemonList from '@Containers/pokemon-list/store/reducers';
import pokemonDetail from '@Containers/pokemon-detail/store/reducers';

const appReducer = combineReducers({
  pokemonList,
  pokemonDetail,
});

export default (state, action) => appReducer(state, action);
