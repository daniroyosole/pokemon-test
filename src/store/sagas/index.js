import { all, takeEvery } from 'redux-saga/effects';

// Container sagas
import { getList } from '@Containers/pokemon-list/store/sagas';
import { getDetail } from '@Containers/pokemon-detail/store/sagas';

import { GET_LIST_REQUEST } from '@Containers/pokemon-list/store/actions/actionTypes';
import { GET_DETAIL_REQUEST } from '@Containers/pokemon-detail/store/actions/actionTypes';

export default function* root() {
  yield all([
    takeEvery(GET_LIST_REQUEST, getList),
    takeEvery(GET_DETAIL_REQUEST, getDetail)
  ]);
};
