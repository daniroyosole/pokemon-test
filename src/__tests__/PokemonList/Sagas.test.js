import { getList, getListSuccess, getListFailure } from '@Containers/pokemon-list/store/sagas';
import "isomorphic-fetch";

test("getList", () => {
    const action = {payload: 1};
    const gen = getList(action);
    gen.next();
    expect(gen.next({results: []}).value).toMatchSnapshot();
    expect(gen.next({results: []}).done).toEqual(true);
});

test("failList", () => {
    const action = {payload: 1};
    const gen = getList(action);
    gen.next();
    expect(gen.next({}).value).toMatchSnapshot();
    expect(gen.next({}).done).toEqual(true);
});
