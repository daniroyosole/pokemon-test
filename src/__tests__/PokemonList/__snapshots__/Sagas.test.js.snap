// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`failList 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "error": [ReferenceError: getListFailure is not defined],
      "type": "GET_LIST_FAILURE",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`getList 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "response": Array [],
      "type": "GET_LIST_SUCCESS",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;
